### PR TITLE
Update add_ml_model and add_script

### PR DESCRIPTION
**Pull request by mellis13**
This PR addresses incorrect or inconsistent behavior in the ``Model.add_ml_model()``, ``Model.add_script()`` and ``Model.add_function()`` API.  The following changes are included in this PR:

- Update documentation to be more accurate
- Correct the code for setting models on multiple GPUs
- Update tests so that they run on GPU and multiple GPUs
- Change the default network interface for test to be ``lo`` instead of ``ipogif0``

### CODE REVIEW

**Review by MattToast — APPROVED** [review]
Two quick requests tangentially related to this PR, but otherwise lgtm!!

**Comment by mellis13** [comment]
Thanks for the great feedback.  I'm going to move this PR to "draft" because I was looking at the code more and see that the tests for these functions were not being tested on GPU for WLM tests when device was set to GPU.  I'll address your feedback and a few other things and change the status back when it is ready for review.

File: smartsim/_core/entrypoints/colocated.py [hunk 1]
```diff
@@ -139,19 +141,15 @@ def launch_db_script(client: Client, db_script: List[str]):
...
-        if args.devices_per_node == 1:
-            client.set_script_from_file(args.name, args.file, args.device)
+        if args.devices_per_node > 1 and args.device.lower() == "gpu":
+            client.set_script_from_file_multigpu(args.name, args.file, 0, args.devices_per_node)
         else:
-            for device_num in range(args.devices_per_node):
-                client.set_script_from_file(
-                    args.name, args.file, args.device + f":{device_num}"
-                )
+            client.set_script_from_file(args.name, args.file, args.device)
```
[the remark this answers is not in the document]
Comment (mellis13) [reply smartsim/_core/entrypoints/colocated.py:153]: Good idea -- fixing

File: tests/backends/test_dbscript.py [hunk 2]
```diff
@@ -47,321 +50,523 @@ def timestwo(x):
...
     colo_model.set_path(test_dir)
-    colo_model.colocate_db(
-        port=wlmutils.get_test_port(),
+    colo_model.colocate_db_tcp(
+        port=test_port,
         db_cpus=1,
         limit_app_cpus=False,
         debug=True,
-        ifname="lo",
+        ifname=test_interface,
```
Comment (MattToast) [inline tests/backends/test_dbscript.py:242]: Same "I imagine `ifname="lo"` to be FAR more popular than `ifname=["ipogif0"]` for colocated dbs. Should we consider parameteriziing this test to continue to explicitly test against `"lo"`?" Thought here and for remaining tests in this file

File: smartsim/_core/config/config.py [hunk 3]
```diff
@@ -158,6 +162,10 @@ def test_launcher(self) -> str:  # pragma: no cover
     def test_device(self) -> str:  # pragma: no cover
         return os.environ.get("SMARTSIM_TEST_DEVICE", "CPU")
 
+    @property
+    def test_num_gpus(self) -> str:  # pragma: no cover
+        return os.environ.get("SMARTSIM_TEST_NUM_GPUS", 1)
```
[the remark this answers is not in the document]
Comment (mellis13) [reply smartsim/_core/config/config.py:167]: Good catch -- I will change to ``int(os.environ.get("thing") or 123)`` and also update ``wlm_trials`` and ``jm_interval``

File: tests/backends/test_dbmodel.py [hunk 4]
```diff
@@ -286,319 +350,419 @@ def test_db_model_ensemble(fileutils, wlmutils):
...
 
+    # Set experiment name
     exp_name = "test-colocated-db-model-pytorch"
-    exp = Experiment(exp_name, launcher="local")
 
-    # get test setup
+    # Retrieve parameters from testing environment
+    test_launcher = wlmutils.get_test_launcher()
+    test_interface = "lo"
+    test_interface = wlmutils.get_test_interface()
```
[the remark this answers is not in the document]
Comment (mellis13) [reply tests/backends/test_dbmodel.py:440]: After discussion, we will use ``get_test_interface`` for all of them but follow up with a ticket to better/consistently handle wanted to test both ``lo`` and a system-specific network interface for colocated.

File: smartsim/entity/model.py [hunk 5]
```diff
@@ -356,14 +356,18 @@ def add_ml_model(
...
         :type model: byte string, optional
         :param model_path: serialized model
         :type model_path: file path to model
-        :param backend: name of the backend (TORCH, TF, TFLITE, ONNX)
-        :type backend: str
         :param device: name of device for execution, defaults to "CPU"
         :type device: str, optional
+        :param devices_per_node: The number of GPU devices available on the host.
+               This parameter only applies to GPU devices and will be ignored if device
+               is specified as GPU.
```
[the remark this answers is not in the document]
Comment (mellis13) [reply smartsim/entity/model.py:369]: Done!

**Review by MattToast — APPROVED** [review]
LGTM!! Thanks for the hard work on this one!!

**Pull request merged**